cpp/ql/src/experimental/Security/CWE/CWE-190/DangerousUseOfTransformationAfterOperation.ql

```diff
@@ -0,0 +1,94 @@
+  n in [0 .. fc.getNumberOfArguments() - 1] and
+  fc.getArgument(n) = exp and
+  result = n
+}
+
+/** Holds if a nonsensical type conversion situation is found. */
+predicate conversionDoneLate(MulExpr mexp, Expr e1, Expr e2) {
+  mexp.getConversion().hasExplicitConversion() and
+  mexp.getConversion() instanceof ParenthesisExpr and
+  mexp.getConversion().getConversion() instanceof CStyleCast and
+  mexp.getConversion().getConversion().getType().getSize() > mexp.getType().getSize() and
+  mexp.getConversion().getConversion().getType().getSize() > e2.getType().getSize() and
+  mexp.getConversion().getConversion().getType().getSize() > e1.getType().getSize() and
+  exists(Expr e0 |
+    e0.(AssignExpr).getRValue() = mexp.getParent*() and
+    e0.(AssignExpr).getLValue().getType().getSize() =
+      mexp.getConversion().getConversion().getType().getSize()
+    or
+    mexp.getEnclosingElement().(ComparisonOperation).hasOperands(mexp, e0) and
+    e0.getType().getSize() = mexp.getConversion().getConversion().getType().getSize()
+    or
+    e0.(FunctionCall)
+        .getTarget()
+        .getParameter(argumentPosition(e0.(FunctionCall), mexp, _))
+        .getType()
+        .getSize() = mexp.getConversion().getConversion().getType().getSize()
+  )
+}
+
+/** Holds if the situation of a possible signed overflow used in pointer arithmetic is found. */
+predicate signSmallerWithEqualSizes(MulExpr mexp, Expr e1, Expr e2) {
+  mexp.getConversion+().getUnderlyingType().getSize() = e1.getUnderlyingType().getSize() and
+  (
+    e2.isConstant() or
+    mexp.getConversion+().getUnderlyingType().getSize() = e2.getUnderlyingType().getSize()
+  ) and
+  mexp.getConversion+().getUnderlyingType().getSize() = e1.getUnderlyingType().getSize() and
+  exists(AssignExpr ae |
+    ae.getRValue() = mexp.getParent*() and
+    ae.getRValue().getUnderlyingType().(IntegralType).isUnsigned() and
+    ae.getLValue().getUnderlyingType().(IntegralType).isSigned() and
+    (
+      not exists(DivExpr de | mexp.getParent*() = de)
+      or
+      exists(DivExpr de, Expr ec |
+        e2.isConstant() and
+        de.hasOperands(mexp.getParent*(), ec) and
+        ec.isConstant() and
+        e2.getValue().toInt() > ec.getValue().toInt()
+      )
+    ) and
+    exists(PointerAddExpr pa |
+      ae.getASuccessor+() = pa and
+      pa.getAnOperand().(VariableAccess).getTarget() = ae.getLValue().(VariableAccess).getTarget()
+    )
+  )
+}
+
+from MulExpr mexp, string msg
+where
+  exists(Expr e1, Expr e2 |
+    mexp.hasOperands(e1, e2) and
+    not e1.isConstant() and
+    not e1.hasConversion() and
+    not e1.hasConversion() and
+    (
+      e2.isConstant() or
+      not e2.hasConversion()
+    ) and
+    (
+      conversionDoneLate(mexp, e1, e2) and
+      msg = "this transformation is applied after multiplication"
+      or
+      signSmallerWithEqualSizes(mexp, e1, e2) and
+      msg = "possible signed overflow followed by offset of the pointer out of bounds"
```

Comment:
We have a [range analysis library](https://codeql.github.com/codeql-standard-libraries/cpp/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll/module.SimpleRangeAnalysis.html) that can rule out cases that obviously doesn't overflow. See how it's used here: https://github.com/github/codeql/blob/main/cpp/ql/src/Likely%20Bugs/Arithmetic/SignedOverflowCheck.ql#L30 (there's only a `convertedExprMightOverflow` predicate that checks for both overflow and underflow at the same time). Would that be useful to rule out false positives in some cases?

Reply:
I am considering this idea.
in my experience of using `Range` there are too many surprises of uncertainty.

in the first part, I hope for a difference in the sizes of the types of arguments and results, in the second time, that the signed result is 1 bit less than the arguments.

I propose to return to this point if other cleaning options are exhausted?

Reply:
> in my experience of using Range there are too many surprises of uncertainty.

I'd be interested in hearing what you mean by this.

> I propose to return to this point if other cleaning options are exhausted?

That's fine.